File: pset1/Modulo01 - C 🔋/credit/credit.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

bool validarLuhn(const char *numero) {
    int len = strlen(numero);
    int soma = 0;
    bool duplicar = false;

    for (int i = len - 1; i >= 0; i--) {
        int digito = numero[i] - '0';

        if (duplicar) {
            digito *= 2;

            if (digito > 9) {
                digito = (digito % 10) + 1;
            }
        }

        soma += digito;
        duplicar = !duplicar;
    }

    return (soma % 10 == 0);
}

bool validarAMEX(const char *numero) {
    int len = strlen(numero);

    if (len != 15 || numero[0] != '3' || (numero[1] != '4' && numero[1] != '7')) {
        return false;
    }

    return validarLuhn(numero);
}

bool validarMASTERCARD(const char *numero) {
    int len = strlen(numero);

    if (len != 16 || numero[0] != '5' || (numero[1] < '1' || numero[1] > '5')) {
        return false;
    }

    return validarLuhn(numero);
}

bool validarVISA(const char *numero) {
    int len = strlen(numero);

    if ((len != 13 && len != 16) || numero[0] != '4') {
        return false;
    }

    return validarLuhn(numero);
}

int main() {
    char numero[20];

    printf("Digite somente números sem espaços do cartão de crédito: ");
    scanf("%s", numero);

    if (validarAMEX(numero)) {
        printf("Número do cartão AMEX válido.\n");
    } else if (validarMASTERCARD(numero)) {
        printf("Número do cartão MASTERCARD válido.\n");
    } else if (validarVISA(numero)) {
        printf("Número do cartão VISA válido.\n");
    } else {
        printf("Número do cartão Inválido.\n");
    }

    return 0;
}

File: pset1/Modulo06 - Python 🐍/credit/credit.py
def validar_luhn(numero):
    len_numero = len(numero)
    soma = 0
    duplicar = False

    for i in range(len_numero - 1, -1, -1):
        digito = int(numero[i])

        if duplicar:
            digito *= 2

            if digito > 9:
                digito = (digito % 10) + 1

        soma += digito
        duplicar = not duplicar

    return soma % 10 == 0


def validar_amex(numero):
    len_numero = len(numero)

    if len_numero != 15 or numero[0] != '3' or (numero[1] != '4' and numero[1] != '7'):
        return False

    return validar_luhn(numero)


def validar_mastercard(numero):
    len_numero = len(numero)

    if len_numero != 16 or numero[0] != '5' or (numero[1] < '1' or numero[1] > '5'):
        return False

    return validar_luhn(numero)


def validar_visa(numero):
    len_numero = len(numero)

    if (len_numero != 13 and len_numero != 16) or numero[0] != '4':
        return False

    return validar_luhn(numero)


numero = input("Digite somente números sem espaços do cartão de crédito: ")

if validar_amex(numero):
    print("Número do cartão AMEX válido.")
elif validar_mastercard(numero):
    print("Número do cartão MASTERCARD válido.")
elif validar_visa(numero):
    print("Número do cartão VISA válido.")
else:
    print("Número do cartão Inválido.")
